OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+                dF2_dfiberVelocityNorm * dfiberVelocityNorm_dprojFibVelNorm;
+        dF2_dprojFibLenNorm =
+                dF2_dfiberVelocityNorm * dfiberVelocityNorm_dprojFibLenNorm;
+    }
+
+
+    //------F3 is the dimensionless fiber (PEE) force (in units of Fmax)------//
+
+
+
+    double dF3_dprojFibLenNorm;
+
+    // stiffness of the linear term is 1 N/m, convert to Fmax/Lceopt units
+    double kPEE = 1.0 / maxIsoForce * optFiberLength;
+    // elongation of fiber (PEE), relative to Lceopt
+    double kPEE2Norm = 1 / pow(fl_width, 2);  //Fiber (PEE) quadratic stiffness,
+    //          so Fpee = Fmax when Lce = Lce*(1+W)
+    double elongationFiberNorm = (fiberLengthNorm - fiberSlackLengthNorm);
+    double F3 = kPEE * elongationFiberNorm;
+    // low stiffness linear term
+    double dF3_dfiberLengthNorm = kPEE;
+    if (elongationFiberNorm > 0) {
+        //add quadratic term for positive elongation
+        F3 = F3 + kPEE2Norm * pow(elongationFiberNorm, 2);
+        if (returnJacobians) {
+            dF3_dfiberLengthNorm =
+                    dF3_dfiberLengthNorm + 2 * kPEE2Norm * elongationFiberNorm;
+        }
+    }
+    if (returnJacobians) {
+        dF3_dprojFibLenNorm = dF3_dfiberLengthNorm * dfiberLength_dprojFibLen;
+    }
+
+
+    //--------F4 is the dimensionless SEE force (in units of Fmax)----------//
+
+
+    // stiffness of the linear term is 1 N/m, convert to Fmax/m (so normalized
+    // by Fmax)
+    double kSEE = 1.0 / maxIsoForce;
+
+    // Tendon (SEE) quadratic stiffness, so Fsee = Fmax at strain of umax
+    // This is normalized by Fmax
+    double kSEE2 = maxIsoForce / pow(tendonSlackLength * fMaxTendonStrain, 2);
+
+    // elongation of tendon (SEE), in meters
+    double elongationTendon = muscleLength - projFibLenNorm *
+                                             optFiberLength - tendonSlackLength;
+    //  low stiffness linear term
+    double F4 = kSEE * elongationTendon;
+    double dF4_dmuscleLength;
+    double dF4_dprojFibLenNorm;
+    if (returnJacobians) {
+         dF4_dprojFibLenNorm = -kSEE * optFiberLength;
+         dF4_dmuscleLength = kSEE;
+    }
+    if (elongationTendon > 0) {
+        // add quadratic term for positive deformation
+        F4 = F4 + (kSEE2 * pow(elongationTendon, 2))/maxIsoForce;
+        if (returnJacobians) {
+             dF4_dprojFibLenNorm = dF4_dprojFibLenNorm - 2 * kSEE2 *
+                                             optFiberLength* elongationTendon;
+             dF4_dmuscleLength = dF4_dmuscleLength + 2 * kSEE2 *
+                                                             elongationTendon;
+        }
+    }
+
+
+    //-- F5 is viscous damping parallel to the CE (0.001 of Fmax at 1 Lceopt/s)
+    // to  ensure that df/dLcedot is never zero-----------//
+    double F5 = dampingCoeff * projFibVelNorm ;
+    //TODO:  Should this be: dampingCoeff * fiberVelocity  (not proj)?
+    //TODO:  Why are damping coeff used instead of damping ratio?
```

Comment:
> Should this be: dampingCoeff * fiberVelocity  (not proj)?
I need to talk to Ton about this.  The problem that I have is that as pennation goes to 90, the damping disappears.

> Why are damping coeff used instead of damping ratio?
To me damping ratio would be better, but then you need muscle effective "mass" to calculate "b" (coeff) from a user provided "z" (ratio). That may get complicated.  It would be better for robustness though.  since it is not a parameter that (I believe) has a strong basis in literature and there for numerical reasons, we should maybe think on it.